include/pybind11/factory.h

```diff
@@ -0,0 +1,269 @@
+/*
+    pybind11/factory.h: Helper class for binding C++ factory functions
+                        as Python constructors.
+
+    Copyright (c) 2017 Jason Rhinelander <jason@imaginary.ca>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+#include "pybind11.h"
+
+#if defined(_MSC_VER)
+#  pragma warning(push)
+#  pragma warning(disable: 4100) // warning C4100: Unreferenced formal parameter
+#  pragma warning(disable: 4127) // warning C4127: Conditional expression is constant
+#endif
+
+NAMESPACE_BEGIN(pybind11)
+
+template <typename type, typename... options>
+template <typename... Args, typename... Extra>
+class_<type, options...> &class_<type, options...>::def(detail::init_factory<Args...> &&init, const Extra&... extra) {
+    std::move(init).execute(*this, extra...);
+    return *this;
+}
+
+NAMESPACE_BEGIN(detail)
+
+template <typename Func, typename Return, typename... Args> struct init_factory {
+private:
+    using FuncType = typename std::remove_reference<Func>::type;
+    using PlainReturn = typename std::remove_pointer<Return>::type;
+    using ForwardReturn = typename std::add_rvalue_reference<Return>::type;
+
+    template <typename Class> using Cpp = typename Class::type;
+    template <typename Class> using Alias = typename Class::type_alias;
+    template <typename Class> using Inst = typename Class::instance_type;
+    template <typename Class> using Holder = typename Class::holder_type;
+
+    template <typename Class, typename SFINAE = void> struct alias_constructible : std::false_type {};
+    template <typename Class> struct alias_constructible<Class, enable_if_t<Class::has_alias &&
+            std::is_convertible<Return, Alias<Class>>::value && std::is_constructible<Alias<Class>, Alias<Class> &&>::value>>
```

Comment:
`std::is_constructible<Alias<Class>, Alias<Class> &&>` == `std::is_move_constructible<Alias<Class>>` which might better reveal the intent.